Add queryElements test helper, remove t.plan

smhigley opened the pull request:
Resolves #1312 

Adds `t.context.queryElements` helper function that wraps `t.context.session.findElements`, and fails the test if no elements are found.

Removes `t.plan()`, now that it isn't needed to ensure that tests are run.

Removes some unrelated unused imports caught while updating the above.

zcorpan commented on test/util/queryElements.js:
```diff
@@ -0,0 +1,23 @@
+  }
+  else {
+    return result;
+  }
```
`result` can be returned unconditionally, right?
```suggestion
  }
  return result;
```

zcorpan commented on test/util/queryElements.js:
```diff
@@ -0,0 +1,23 @@
+/**
+ * Return an array of elements by selector. Wraps Selenium's findElements, but with a failing text for empty queries
+ *
+ * @param {String} selector - CSS selector string
```
`t` is not documented

zcorpan commented:
I added a lint check in c16088f -- there were still 12 uses (not counting the one in `queryElement.js`). Do you think those should be changed to use `queryElements`?

```
/Users/zcorpan/git/w3c/aria-practices/test/index.js
  89:16  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties

/Users/zcorpan/git/w3c/aria-practices/test/tests/checkbox_checkbox-2.js
   55:14  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties
   85:12  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties
  201:12  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties

/Users/zcorpan/git/w3c/aria-practices/test/tests/combobox_autocomplete-both.js
  60:31  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties

/Users/zcorpan/git/w3c/aria-practices/test/tests/combobox_autocomplete-list.js
  61:31  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties

/Users/zcorpan/git/w3c/aria-practices/test/tests/combobox_autocomplete-none.js
  61:31  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties

/Users/zcorpan/git/w3c/aria-practices/test/tests/combobox_grid-combo.js
  137:31  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties

/Users/zcorpan/git/w3c/aria-practices/test/tests/grid_LayoutGrids.js
  125:16  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties

/Users/zcorpan/git/w3c/aria-practices/test/tests/tabs_tabs-1.js
  355:12  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties

/Users/zcorpan/git/w3c/aria-practices/test/tests/tabs_tabs-2.js
  363:12  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties

/Users/zcorpan/git/w3c/aria-practices/test/util/assertAriaDescribedby.js
  51:36  error  'findElements' is restricted from being used. Please use t.context.queryElements()  no-restricted-properties

✖ 12 problems (12 errors, 0 warnings)
```

zcorpan commented:
Documentation in the wiki should be updated. I don't want to do this now since my changes haven't been reviewed and I don't want the documentation to be out of sync with what's in `master`.

* [ ] The eslint for `findElements`, and the per-file override `/* eslint no-restricted-properties: 0 */`
* [ ] [How to declare a test](https://github.com/w3c/aria-practices/wiki/Introduction-to-APG-Regression-Tests#how-to-declare-a-test)
* [ ] [Example test](https://github.com/w3c/aria-practices/wiki/Regression-Tests-for-APG-Example-Pages#example-test)
* [ ] [Selenium API](https://github.com/w3c/aria-practices/wiki/Regression-Tests-for-APG-Example-Pages#selenium-api)

mcking65 commented:
@smhigley, could you please look at Simon's changes?

It would be good to have a review of the entire PR from @sh0ji or @spectranaut as well.

Given the number of files touched, I'd like to get this merged as soon as we can.

nschonni commented on .eslintrc.json:
```diff
@@ -5,6 +5,10 @@
     "no-unused-vars": 0,
-    "no-undef": 0
+    "no-undef": 0,
+    "no-restricted-properties": [2, {
```
Should `plan` be added if you're trying to discourage it too?

nschonni commented on test/index.js:
```diff
@@ -1,3 +1,5 @@
+/* eslint no-restricted-properties: 0 */
```
Might makes sense to use a single line ignore instead of a file level one

nschonni replied:
Looks like it can be re-enabled now

smhigley commented:
I left those 12 in there, because they dealt with testing the number returned, rather than just assuming within the test that there would be a non-empty array. I think that could still be a valid use case for `findElements` instead of `queryElements`.

What if we add a lint rule for `t.plan` instead, as @nschonni recommended? If that seems like it increases complexity too much for people writing tests, I also like @spectranaut's suggestion of a helper for that specific scenario. Perhaps something like `assertNotEmpty` or `assertNumberOfElements` (I'm bad at naming things... 😝)

zcorpan commented:
I considered linting `plan` but wasn't sure if there are still cases where it's needed. If there aren't then we can lint it.

For the cases where `findElements` is still used, we want to assert that the length is 0? A separate helper function for that which fails if the length is not 0 seems cleanest to me. `assertNoElement()` maybe?

mcking65 commented:
What is the best approach to arriving at a decision for how to resolve the remaining issue here? The technical aspect of this discussion is beyond my expertise.

One thing I hope we optimize for is clarity of code, especially for new comers to test writing.

smhigley commented:
I like @zcorpan's suggestion of `assertNoElements`. I'll push up a commit with it in just a bit :)

smhigley commented:
I rebased and removed all the remaining uses of `findElements`, apart from the two in the `queryElements` and `assertNoElements` helpers.

@spectranaut / @zcorpan / @nschonni, how do you feel about the PR after those changes?

nschonni commented on test/tests/checkbox_checkbox-2.js:
```diff
@@ -1,10 +1,13 @@
+/* eslint no-restricted-properties: 0 */
```
Think this can be re-enabled again

smhigley commented:
@nschonni I removed the now-unnecessary `eslint no-restricted-properties`. `eslint . --fix` doesn't seem to be catching or changing anything, though

nschonni commented:
@smhigley don't worry about it then :)
Looks like some mixed 2/4 spaces in some functions, but if the lint isn't catching it then the rules aren't enforcing it right now and shouldn't matter here

smhigley commented:
@spectranaut are you ok with the PR now/do you have any outstanding concerns?

zcorpan commented:
(I didn't get to resolving conflicts today.)

smhigley commented:
@zcorpan conflicts resolved :)
